pkg/types/alertmanagerconfig.go

```diff
@@ -0,0 +1,126 @@
+package alertmanagerconfig
+
+import (
+	"fmt"
+
+	yaml "gopkg.in/yaml.v2"
+)
+
+// The following types are taken from the upstream Alertmanager types, and modified
+// to allow printing of Secrets so that we can generate valid configs from them.
+// The Alertmanager types are not supported as external libraries, and therefore need
+// to be recreated for this operator.
+// Discussion, for reference, is in this PR: https://github.com/prometheus/alertmanager/pull/1804
+
+type Config struct {
+	Global    *GlobalConfig `yaml:"global,omitempty" json:"global,omitempty"`
+	Route     *Route        `yaml:"route,omitempty" json:"route,omitempty"`
+	Receivers []*Receiver   `yaml:"receivers,omitempty" json:"receivers,omitempty"`
+	Templates []string      `yaml:"templates" json:"templates"`
+
+	// original is the input from which the config was parsed.
+	original string
+}
+
+func (c Config) String() string {
+	b, err := yaml.Marshal(c)
+	if err != nil {
+		return fmt.Sprintf("<error creating config string: %s>", err)
+	}
+	return string(b)
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface for Config.
+func (c *Config) UnmarshalYAML(unmarshal func(interface{}) error) error {
```

Comment:
Why are you implementing the unmarshal function instead of using the built in ?

Reply:
This is something I copied from the Alertmanager Config data type, from an upstream repo. If you unmarshal an Alertmanager Config using the default unmarshaler, it gives you a bunch of memory addresses instead of valid yaml. For example, this is how the list of receivers is printed out:

```
[0xc4201831d0 0xc4201832c0 0xc4201833b0]
```

And individual receivers look like this:
```
&{pagerduty [] [0xc0009220e0] [] [] [] [] [] [] []}
```

It also helps to ensure that the list of receivers each has a unique name, which is required for a valid Alertmanager config.